### ISSUE TITLE
grep issues

### ISSUE THREAD

**Issue by redguardtoo** [abo-abo/swiper#1342]
`counsel-grep` does not work 
`counsel-git-grep`, `counsel-rg`, and `counsel-ag` fail when input negative keywords ("keyword1 !keyword2", for example).

Tested on ivy `v0.10.0`

**Comment by basil-conto** [comment]
> `counsel-grep` does not work

Please elaborate on the issue with steps to reproduce, etc. as that gives nothing to go on.

**Comment by basil-conto** [comment]
The other commands currently hardcode the `ivy--regex` regexp builder; perhaps it's possible to lift this restriction, but keep in mind that there will always be discrepancies between Emacs and external regexp syntax to some extent.

**Comment by abo-abo** [comment]
> there will always be discrepancies between Emacs and external regexp syntax to some extent

Most external tools don't support negative patterns. We could fix this by either doing the negative match filtering in Emacs, or tailoring each shell command by appending `grep -v`.

**Comment by basil-conto** [comment]
@abo-abo Indeed, that's what I meant by "perhaps it's possible to lift this restriction", i.e. allow more flexible interaction with external utilities. My comment about the regexp syntax discrepancies was more theoretical / philosophical and, thus, of little use. :)

**Comment by redguardtoo** [comment]
@abo-abo , we can use both solution, if `grep` exists use it, else use elisp.

**Comment by basil-conto** [comment]
@redguardtoo 

> if `grep` exists use it

Isn't that the whole point of `counsel-grep` vs, say, `counsel-grep-or-swiper` or `swiper`?

**Comment by redguardtoo** [comment]
@basil-conto . Thanks for the heads up, here is my new set up of `counsel-git-grep`

```
(defun counsel-etags-positive-regex (patterns)
  "Extract positive regex from PATTERNS."
  (let* ((re (car patterns)))
    (cond
     ((or (not re) (string= re ""))
      "[^ \t]+")
     (t
      (ivy--regex re)))))

(defun counsel-etags-exclusion-regex (patterns)
  "Extract exclusion PATTERNS."
  (let* ((re (cadr patterns)))
    (unless re (setq re ""))
    ;; remove trailing spaces
    (setq re (replace-regexp-in-string " +$" "" re))
    (cond
     ((string= re "")
      (setq re nil))
     (t
      (mapconcat 'ivy--regex
                 (split-string re " +")
                 "\\\|")))))

(with-eval-after-load 'counsel

  (defun my-counsel-git-grep-cmd-function-default (str)
    (let* ((patterns (split-string str " *!"))
           (pos-re (counsel-etags-positive-regex patterns))
           (neg-re (counsel-etags-exclusion-regex patterns))
           rlt)
      (setq rlt (format counsel-git-grep-cmd pos-re))
      (when neg-re
        (setq rlt (format "%s --and --not -e \"%s\"" rlt neg-re)))
     rlt))

  (setq counsel-git-grep-cmd-function #'my-counsel-git-grep-cmd-function-default))

```

**Comment by basil-conto** [comment]
> Thanks for the heads up, here is my new set up of `counsel-git-grep`

As another heads up, I'm slowly working on a patch that changes the default value of `counsel-git-grep-cmd-default` to
```el
(list "git" "--no-pager"
      "grep" "--line-number" "--no-color" "-I" "%s")
```
(notice the removal of `-e` by default).

This will allow `counsel-git-grep-cmd-function-ignore-order` to add `-e`, `--and`, `--not`, etc. as needed, without having to fiddle with extraneous `-e` flags.
